Support payer-sponsored auto-creation with HTS assets (#542)
hashgraph/hedera-improvement-proposal: HIP/hip-000-hts-asset-auto-creation.md

that sends hbar to a protobuf representation of a key called an _alias_.

When this alias has not already been used `CryptoTransfer`, **and** the sent hbar is enough to cover the account creation 
fee, then the network creates a new account with the key implied by the alias. It deducts the account creation fee from 
the sent hbar, and uses whatever is left over as the new account's balance. 

Thus auto-account creation does not work if sending only a units of a fungible token or an NFT to a new alias, 
because an hbar fee cannot be deducted from HTS assets. 

We propose to support auto-creation via HTS assets by charging the account creation fee to the payer of the triggering 
`CryptoTransfer`. This means a new alias may be given anywhere in the transaction; both in the hbar transfer list, or in 
an HTS token transfer list. In the latter case, the assessed creation fee will include at least one auto-association slot, 
since the new account must be associated to its originating HTS assets.

**IMPORTANT:** We do _not_ propose to make nodes do extra work in precheck estimating implied auto-creation fees. These

> fees will be assessed as part of the `CryptoTransfer` service logic; and, as with all other service fees, will not be 

Reviewer: ```suggestion
fees will be assessed as part of the `CryptoTransfer` service logic; as with all other service fees, they will not be 
```

estimated in precheck.

## Motivation

> Allowing auto-creation with token assets makes the network match the expectations of Ethereum users, since Ethereum 

Reviewer: ```suggestion
Allowing account auto-creation with token assets makes the network match the expectations of Ethereum users since Ethereum 
```

> contracts allow ERC-20 and ERC-721 assets be transferred to an externally operated account (EOA) even _before_ that 

Reviewer: ```suggestion
contracts enable transferring ERC-20 and ERC-721 assets to an externally operated account (EOA) even _before_ that 
```

account has been created by sending it value. 

## Rationale

> We cannot provide free account creation, for the same DOS risks outlined in HIP-32. There is also no reasonable way to 

Reviewer: ```suggestion
We cannot provide free account creation for the same DOS risks outlined in HIP-32. However, there is also no reasonable way to 
```

> convert HTS token units or NFTs into hbar. It follows that the only solution to extend `CryptoTransfer` auto-creation to 

Reviewer: ```suggestion
convert HTS token units or NFTs into hbar. Therefore, a solution to extend `CryptoTransfer` auto-creation to 
```

HTS assets is to deduct the account creation fee from the `CryptoTransfer` payer.

## User stories

- As an NFT creator, I want to send an NFT to an onboarding collector, given only their public key. 
- As a dApp operator, I want to send my native token to an onboarding user in a single transaction; rather
than first submitting a `CryptoCreate`, then a `TokenAssociate`, then a `CryptoTransfer`.
  
## Specification

The technical specification for this HIP requires three changes in the network:

>  - Instead of returning `INVALID_ACCOUNT_ID` when a new alias is credited with a non-hbar asset,

Reviewer: Passive voice

```suggestion
 - Instead of returning `INVALID_ACCOUNT_ID` when crediting a new alias with a non-hbar asset,
```

> run the same auto-creation logic that runs today when a new alias is credited with hbar.

Reviewer: Passive voice

```suggestion
run the same auto-creation logic that runs today when crediting a new alias with hbar.
```

 - Instead of deducting the account creation fee from the assets being transferred to a new 
alias, charge this creation fee to the `CryptoTransfer` payer account.

>  - When a new alias is receiving _n_ different types of HTS assets, create the synthetic `CryptoCreate` 

Reviewer: ```suggestion
 - When a new alias receives _n_ different types of HTS assets, create the synthetic `CryptoCreate` 
```

transaction with `maxAutoAssociationSlots=n`.

## Backwards Compatibility

Any existing clients that use auto-creation need to be sure the payer account for the 
`CryptoTransfer` can afford an account creation fee, since this fee will no longer be deducted
from the auto-created account's initial balance. 

## Security Implications

This HIP does not change the charging and throttling protections used to secure auto-creation as added in HIP-32. 
So we do not perceive any new security implications.

## How to Teach This

"Auto-create an account by sending HTS assets to its desired alias."

## Reference Implementation

Please follow [this issue](https://github.com/hashgraph/hedera-services/issues/3763) for progress
on the reference implementation.

## Rejected Ideas

> We briefly considered requiring a simultaneous hbar payment upon auto-creation with an HTS assets; but this would have

Reviewer: ```suggestion
We briefly considered requiring a simultaneous hbar payment upon auto-creation with HTS assets, but this would have
```
